storagenode/gracefulexit/chore_test.go

```diff
@@ -4,52 +4,133 @@
 package gracefulexit_test
 
 import (
+	"context"
 	"testing"
 	"time"
 
 	"github.com/stretchr/testify/require"
 
+	"storj.io/storj/internal/memory"
 	"storj.io/storj/internal/testcontext"
 	"storj.io/storj/internal/testplanet"
+	"storj.io/storj/internal/testrand"
+	"storj.io/storj/pkg/storj"
 	"storj.io/storj/satellite/overlay"
+	"storj.io/storj/storagenode"
+	"storj.io/storj/uplink"
 )
 
 func TestChore(t *testing.T) {
 	testplanet.Run(t, testplanet.Config{
 		SatelliteCount:   1,
-		StorageNodeCount: 8,
+		StorageNodeCount: 9,
 		UplinkCount:      1,
 	}, func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet) {
 		satellite1 := planet.Satellites[0]
-		exitingNode := planet.StorageNodes[0]
+		uplinkPeer := planet.Uplinks[0]
 
 		satellite1.GracefulExit.Chore.Loop.Pause()
+
+		rs := &uplink.RSConfig{
+			MinThreshold:     4,
+			RepairThreshold:  6,
+			SuccessThreshold: 8,
+			MaxThreshold:     8,
+		}
+
+		err := uplinkPeer.UploadWithConfig(ctx, satellite1, rs, "testbucket", "test/path1", testrand.Bytes(5*memory.KiB))
+		require.NoError(t, err)
+
+		exitingNode, err := findNodeToExit(ctx, planet, 1)
+		require.NoError(t, err)
 		exitingNode.GracefulExit.Chore.Loop.Pause()
 
 		exitStatus := overlay.ExitStatusRequest{
 			NodeID:          exitingNode.ID(),
 			ExitInitiatedAt: time.Now(),
 		}
 
-		_, err := satellite1.Overlay.DB.UpdateExitStatus(ctx, &exitStatus)
+		_, err = satellite1.Overlay.DB.UpdateExitStatus(ctx, &exitStatus)
 		require.NoError(t, err)
 
 		err = exitingNode.DB.Satellites().InitiateGracefulExit(ctx, satellite1.ID(), time.Now(), 10000)
 		require.NoError(t, err)
 
+		// check that the storage node is exiting
 		exitProgress, err := exitingNode.DB.Satellites().ListGracefulExits(ctx)
 		require.NoError(t, err)
 		require.Len(t, exitProgress, 1)
 
+		// initiate graceful exit on satellite side by running the SN chore.
 		exitingNode.GracefulExit.Chore.Loop.TriggerWait()
 
-		exitProgress, err = exitingNode.DB.Satellites().ListGracefulExits(ctx)
+		// run the satellite chore to build the transfer queue.
+		satellite1.GracefulExit.Chore.Loop.TriggerWait()
+
+		// check that the satellite knows the storage node is exiting.
+		exitingNodes, err := satellite1.DB.OverlayCache().GetExitingNodes(ctx)
+		require.NoError(t, err)
+		require.Len(t, exitingNodes, 1)
+
+		queueItems, err := satellite1.DB.GracefulExit().GetIncomplete(ctx, exitStatus.NodeID, 10, 0)
 		require.NoError(t, err)
+		require.Len(t, queueItems, 1)
 
+		// run the SN chore again to start processing transfers.
+		exitingNode.GracefulExit.Chore.Loop.TriggerWait()
+
+		exitProgress, err = exitingNode.DB.Satellites().ListGracefulExits(ctx)
+		require.NoError(t, err)
 		for _, progress := range exitProgress {
 			if progress.SatelliteID == satellite1.ID() {
 				require.NotNil(t, progress.FinishedAt)
 			}
 		}
 	})
```

Comment:
Maybe we should also check to ensure that the piece that needed to be transferred has shown up on one of the remaining 8 nodes.

Reply:
We can iterate over metainfo before the transfer(s), then keep track of all the paths/piece numbers associated with the exiting node. Then, after the transfers, we can check metainfo for the same segments, and expect that the same piece numbers are still in the pointers, but are associated with different nodes.

Reply:
Updated